cpp/src/text/minhash.cu

```diff
@@ -51,79 +50,96 @@ namespace {
+ *
+ * @param d_strings Strings column to process
+ * @param seeds Seeds for hashing each string
+ * @param width Substring window size in characters
+ * @param d_hashes Minhash output values for each string
  */
-struct minhash_fn {
-  cudf::column_device_view d_strings;
-  cudf::device_span<cudf::hash_value_type const> seeds;
-  cudf::size_type width;
-  cudf::hash_value_type* d_hashes;
-
-  __device__ void operator()(std::size_t idx)
-  {
-    auto const str_idx  = static_cast<cudf::size_type>(idx / cudf::detail::warp_size);
-    auto const lane_idx = static_cast<cudf::size_type>(idx % cudf::detail::warp_size);
-
-    if (d_strings.is_null(str_idx)) { return; }
-
-    auto const d_str    = d_strings.element<cudf::string_view>(str_idx);
-    auto const d_output = d_hashes + (str_idx * seeds.size());
-
-    // initialize hashes output for this string
-    if (lane_idx == 0) {
-      auto const init = d_str.empty() ? 0 : std::numeric_limits<cudf::hash_value_type>::max();
-      thrust::fill(thrust::seq, d_output, d_output + seeds.size(), init);
-    }
-    __syncwarp();
-
-    auto const begin = d_str.data() + lane_idx;
-    auto const end   = d_str.data() + d_str.size_bytes();
-
-    // each lane hashes 'width'  substrings of d_str
-    for (auto itr = begin; itr < end; itr += cudf::detail::warp_size) {
-      if (cudf::strings::detail::is_utf8_continuation_char(*itr)) { continue; }
-      auto const check_str =  // used for counting 'width' characters
-        cudf::string_view(itr, static_cast<cudf::size_type>(thrust::distance(itr, end)));
-      auto const [bytes, left] =
-        cudf::strings::detail::bytes_to_character_position(check_str, width);
-      if ((itr != d_str.data()) && (left > 0)) { continue; }  // true if past the end of the string
-
-      auto const hash_str = cudf::string_view(itr, bytes);
-      // hashing with each seed on the same section of the string is 10x faster than
-      // computing the substrings for each seed
-      for (std::size_t seed_idx = 0; seed_idx < seeds.size(); ++seed_idx) {
-        auto const hasher =
-          cudf::hashing::detail::MurmurHash3_x86_32<cudf::string_view>{seeds[seed_idx]};
+template <
+  typename HashFunction,
+  typename hash_value_type = std::
+    conditional_t<std::is_same_v<typename HashFunction::result_type, uint32_t>, uint32_t, uint64_t>>
+__global__ void minhash_kernel(cudf::column_device_view const d_strings,
+                               cudf::device_span<hash_value_type const> seeds,
+                               cudf::size_type width,
+                               hash_value_type* d_hashes)
+{
+  auto const idx = static_cast<std::size_t>(threadIdx.x + blockIdx.x * blockDim.x);
+  if (idx >= (static_cast<std::size_t>(d_strings.size() *
+                                       static_cast<std::size_t>(cudf::detail::warp_size)))) {
+    return;
+  }
+
+  auto const str_idx  = static_cast<cudf::size_type>(idx / cudf::detail::warp_size);
```

Comment:
In light of the recent discussions about #10368, do we need to be worried about this ever overflowing? Specifically since it's coming from `idx`, which is a `std::size_t`?

Reply:
Good question. No, the `str_idx` should always be within `size_type` after this division otherwise we would have other issues. This is consistent with the recent discussions and changes as well.